contracts/ExponentialStaking.sol

```diff
@@ -0,0 +1,270 @@
+
+        delete lockups[msg.sender][lockupId]; // Keeps empty in array, so indexes are stable
+        _burn(msg.sender, points);
+        if (penalty > 0) {
+            asset.transfer(address(rewardsSource), penalty);
+            emit Penalty(msg.sender, penalty);
+        }
+        asset.transfer(msg.sender, withdrawAmount);
+        emit Unstake(msg.sender, lockupId, withdrawAmount, end, points);
+    }
+
+    // 3. Reward functions
+
+    /// @notice Collect all earned asset rewards.
+    function collectRewards() external {
+        _collectRewards(msg.sender, false);
+    }
+
+    /// @dev Internal function to handle rewards accounting.
+    ///
+    /// 1. Collect new rewards for everyone
+    /// 2. Calculate this user's rewards and accounting
+    /// 3. Distribute this user's rewards
+    ///
+    /// This function *must* be called before any user balance changes.
+    ///
+    /// This will always update the user's rewardDebtPerShare to match
+    /// accRewardPerShare, which is essential to the accounting.
+    ///
+    /// @param user to collect rewards for
+    /// @param shouldRetainRewards if true user's rewards kept in this contract rather than sent
+    /// @return retainedRewards amount of rewards not sent to user
+    function _collectRewards(address user, bool shouldRetainRewards) internal returns (uint256) {
+        uint256 supply = totalSupply();
+        if (supply > 0) {
+            uint256 preBalance = asset.balanceOf(address(this));
+            try rewardsSource.collectRewards() {}
+            catch {
+                // Governance staking should continue, even if rewards fail
+            }
+            uint256 collected = asset.balanceOf(address(this)) - preBalance;
+            accRewardPerShare += (collected * 1e12) / supply;
+        }
+        uint256 netRewardsPerShare = accRewardPerShare - rewardDebtPerShare[user];
+        uint256 netRewards = (balanceOf(user) * netRewardsPerShare) / 1e12;
+        rewardDebtPerShare[user] = accRewardPerShare;
+        if (netRewards == 0) {
+            return 0;
+        }
+        emit Reward(user, netRewards);
+        if (shouldRetainRewards) {
+            return netRewards;
+        } else {
+            asset.transfer(user, netRewards);
+        }
+    }
+
+    /// @notice Preview the number of points that would be returned for the
+    /// given amount and duration.
+    ///
+    /// @param amount asset to be staked
+    /// @param duration number of seconds to stake for
+    /// @return points staking points that would be returned
+    /// @return end staking period end date
+    function previewPoints(uint256 amount, uint256 duration) public view returns (uint256, uint256) {
+        require(duration <= maxStakeDuration, "Staking: Too long");
+        uint256 start = block.timestamp > epoch ? block.timestamp : epoch;
+        uint256 end = start + duration;
+        uint256 endYearpoc = ((end - epoch) * 1e18) / 365 days;
```

Comment:
Can you explain this part of the code?

Reply:
There's basically a prestaking phase where all stakes are equal until staking officially begins.

So the start date of a stake made during prestaking is the official start of all staking (epoch)

The end date is calculated based on the duration from this start date.

To calculate the multiplier that the staking points get, we need to know many years the end date of the stake is from the start of staking.

Reply:
What I am not sure to understand is since `maxStakeDuration` is 365 days, isn't everyone going to have the same year?

Reply:
Let's say:

Staking start day is June 4th, 2024

Current day is May 5th. User stakes for one year. The start date gets pushed up to June 4th, 2024. The end date is approximately June 5th, 2024. 

Then end date - staking start day is 1.0 years, which means that 1.4**1.0 = the user will get 1.4 staking points per OGN staked.

Reply:
The "years" are 1e18 fixed point decimals, and so users are getting a different "year" for every different end timestamp.

Reply:
ok understood.